### DISCUSSION TITLE
Casting issue

### DISCUSSION THREAD

**Discussion by domcir** [bytedeco/javacpp#678] [Q&A] [answered]
Hey

I was able to define a class "CreditDefaultSwap" 
![image](https://github.com/bytedeco/javacpp/assets/54353641/9add29b6-0cf7-41ee-a14a-4cf6355ac5a2)
and configure smart pointers.
![image](https://github.com/bytedeco/javacpp/assets/54353641/19bc9a56-b254-4a91-b8af-ab7a978c6a1e)

That class is later usd as smart pointer argument.
![image](https://github.com/bytedeco/javacpp/assets/54353641/9ffc7f6f-0d77-4f03-bdf9-fe0647d4c6a4)

Unfortunatelly I get a c++ compilation error:
![image](https://github.com/bytedeco/javacpp/assets/54353641/1827f230-e886-4a48-8a70-23b6f4e7ac9f)
so the type isn't recognised on the C++ side as a smart pointer.
`cannot convert argument 1 from 'SharedPtrAdapter<const instruments::CreditDefaultSwap>' to 'const instruments::CreditDefaultSwap &' `.

I can't figure out why?

Thanks
Dom

**Comment by saudet** [comment]
Remove `instruments::CreditDefaultSwap` and `CreditDefaultSwap` from the Info list with annotations. Those are not `std::shared_ptr` so they should not be annotated with `@SharedPtr`.

**Reply by domcir** [reply]
I have only instruments::CreditDefaultSwapPtr and CreditDefaultSwapPtr or instruments::CreditDefaultSwapCPtr and CreditDefaultSwapCPtr and all are just redefinitions of smart pointers so I'm nto sure what you're suggestign to remove.

**Reply by saudet** [reply]
I see, it's just a typedef. So, we probably just need to figure out the right cast, like what happens if you remove the const from the cast?

**Reply by domcir** [reply]
I think I found the issue.

My CDSBootstrapHelper helper inherites from a class whcih is virtual so when i created a defitniion I used:
![image](https://github.com/bytedeco/javacpp/assets/54353641/421dbda5-4c46-4350-b327-f7898af69755)

as a result the logic [here](https://github.com/bytedeco/javacpp/blob/master/src/main/java/org/bytedeco/javacpp/tools/Generator.java#L1595) creates some kind of wrapper in the jni c++ where types are mixed up.

Does it make any sense?

Thanks
Dom

**Reply by saudet** [reply]
Ah, right, adapters don't work too well with virtual functions, that might not work. We can use another cast for return values as the third element of a `@Cast`, that can help get things working.

**Reply by saudet** [reply]
Something like this here: https://github.com/tensorflow/java/pull/345#issuecomment-869085913

**Reply by domcir** [reply]
I can't get it working.
Should I add `valueTypes` with a cast to the definition of the smart pointer?

![image](https://github.com/bytedeco/javacpp/assets/54353641/dee3c382-dd29-4695-b91a-536faaf02687)

but I get this from one of the methods before I even get to the C++ compilation step.
`dlib.java:6267: error: Cast is not a repeatable annotation type
                public native @Cast("const nstruments::CreditDefaultSwap*") @SharedPtr @NoException(true) @Cast({"", "", "instruments::CreditDefaultSwapCPtr"}) CreditDefaultSwap cds();`

I don't understand where to include casting to apply it only to jobject to C++.

**Reply by saudet** [reply]
If you could provide the JNI code that is failing, I might have an idea about what we can do. Can you do that?

**Reply by domcir** [reply]
```cpp
class JavaCPP_hidden JavaCPP__0003a_0003aCTP_0003a_0003amodels_0003a_0003aFwdCDSBootstrapHelper : public ::CTP::models::FwdCDSBootstrapHelper {
public:
    jobject obj;
    static jmethodID isConstant__;

    JavaCPP__0003a_0003aCTP_0003a_0003amodels_0003a_0003aFwdCDSBootstrapHelper(const CTP::instruments::CreditDefaultSwap& arg0, ::CTP::utils::GregorianDateBase<CTP::utils::MicroSecondsTag> arg1, ::CTP::utils::GregorianDateBase<CTP::utils::MicroSecondsTag> arg2, const CTP::termstructures::IRates& arg3, double arg4) : ::CTP::models::FwdCDSBootstrapHelper(arg0, arg1, arg2, arg3, arg4), obj(NULL) { }
    JavaCPP__0003a_0003aCTP_0003a_0003amodels_0003a_0003aFwdCDSBootstrapHelper(const CTP::instruments::CreditDefaultSwap& arg0, ::CTP::utils::GregorianDateBase<CTP::utils::MicroSecondsTag> arg1, ::CTP::utils::GregorianDateBase<CTP::utils::MicroSecondsTag> arg2, const CTP::termstructures::IRates& arg3) : ::CTP::models::FwdCDSBootstrapHelper(arg0, arg1, arg2, arg3), obj(NULL) { }
    using ::CTP::models::FwdCDSBootstrapHelper::isConstant;
    virtual bool isConstant() const noexcept JavaCPP_override;
    bool super_isConstant() const noexcept { return ::CTP::models::FwdCDSBootstrapHelper::isConstant(); }
};
jmethodID JavaCPP__0003a_0003aCTP_0003a_0003amodels_0003a_0003aFwdCDSBootstrapHelper::isConstant__ = NULL;
```

**Reply by domcir** [reply]
the formatting is broken a bit but the issue is that 'const CTP::instruments::CreditDefaultSwap& arg0' should be a reference to smart pointer not a reference to object.

**Reply by domcir** [reply]
The JNI code looks as follows:
```cpp
JNIEXPORT void JNICALL Java_capstone_dlib_00024FwdCDSBootstrapHelper_allocate__Lcapstone_dlib_00024CreditDefaultSwap_2Lcapstone_dlib_00024GregorianDate_2Lcapstone_dlib_00024GregorianDate_2Lcapstone_dlib_00024IRates_2D(JNIEnv* env, jobject obj, jobject arg0, jobject arg1, jobject arg2, jobject arg3, jdouble arg4) {
    ::CTP::instruments::CreditDefaultSwap* ptr0 = arg0 == NULL ? NULL : (::CTP::instruments::CreditDefaultSwap*)jlong_to_ptr(env->GetLongField(arg0, JavaCPP_addressFID));
    if (ptr0 == NULL) {
        env->ThrowNew(JavaCPP_getClass(env, 7), "Pointer address of argument 0 is NULL.");
        return;
    }
    jlong size0 = arg0 == NULL ? 0 : env->GetLongField(arg0, JavaCPP_limitFID);
    void* owner0 = JavaCPP_getPointerOwner(env, arg0);
    jlong position0 = arg0 == NULL ? 0 : env->GetLongField(arg0, JavaCPP_positionFID);
    ptr0 += position0;
    size0 -= position0;
    SharedPtrAdapter< const CTP::instruments::CreditDefaultSwap > adapter0((const CTP::instruments::CreditDefaultSwap*)ptr0, size0, owner0);
    ::CTP::utils::GregorianDateBase<CTP::utils::MicroSecondsTag>* ptr1 = arg1 == NULL ? NULL : (::CTP::utils::GregorianDateBase<CTP::utils::MicroSecondsTag>*)jlong_to_ptr(env->GetLongField(arg1, JavaCPP_addressFID));
    if (ptr1 == NULL) {
        env->ThrowNew(JavaCPP_getClass(env, 7), "Pointer address of argument 1 is NULL.");
        return;
    }
    jlong position1 = arg1 == NULL ? 0 : env->GetLongField(arg1, JavaCPP_positionFID);
    ptr1 += position1;
    ::CTP::utils::GregorianDateBase<CTP::utils::MicroSecondsTag>* ptr2 = arg2 == NULL ? NULL : (::CTP::utils::GregorianDateBase<CTP::utils::MicroSecondsTag>*)jlong_to_ptr(env->GetLongField(arg2, JavaCPP_addressFID));
    if (ptr2 == NULL) {
        env->ThrowNew(JavaCPP_getClass(env, 7), "Pointer address of argument 2 is NULL.");
        return;
    }
    jlong position2 = arg2 == NULL ? 0 : env->GetLongField(arg2, JavaCPP_positionFID);
    ptr2 += position2;
    ::CTP::termstructures::IRates* ptr3 = arg3 == NULL ? NULL : (::CTP::termstructures::IRates*)jlong_to_ptr(env->GetLongField(arg3, JavaCPP_addressFID));
    if (ptr3 == NULL) {
        env->ThrowNew(JavaCPP_getClass(env, 7), "Pointer address of argument 3 is NULL.");
        return;
    }
    jlong size3 = arg3 == NULL ? 0 : env->GetLongField(arg3, JavaCPP_limitFID);
    void* owner3 = JavaCPP_getPointerOwner(env, arg3);
    jlong position3 = arg3 == NULL ? 0 : env->GetLongField(arg3, JavaCPP_positionFID);
    ptr3 += position3;
    size3 -= position3;
    SharedPtrAdapter< const CTP::termstructures::IRates > adapter3((const CTP::termstructures::IRates*)ptr3, size3, owner3);
    jthrowable exc = NULL;
    try {
        ::CTP::models::FwdCDSBootstrapHelper* rptr = new JavaCPP__0003a_0003aCTP_0003a_0003amodels_0003a_0003aFwdCDSBootstrapHelper(adapter0, *ptr1, *ptr2, adapter3, arg4);
        jlong rcapacity = 1;
        JavaCPP_initPointer(env, obj, rptr, rcapacity, rptr, &JavaCPP_capstone_dlib_00024FwdCDSBootstrapHelper_deallocate);
        ((JavaCPP__0003a_0003aCTP_0003a_0003amodels_0003a_0003aFwdCDSBootstrapHelper*)rptr)->obj = env->NewWeakGlobalRef(obj);
    } catch (...) {
        exc = JavaCPP_handleException(env, 9);
    }

    const CTP::instruments::CreditDefaultSwap* rptr0 = (const CTP::instruments::CreditDefaultSwap*)adapter0;
    jlong rsize0 = (jlong)adapter0.size;
    void* rowner0 = adapter0.owner;
    if (rptr0 != (const CTP::instruments::CreditDefaultSwap*)ptr0) {
        JavaCPP_initPointer(env, arg0, rptr0, rsize0, rowner0, &SharedPtrAdapter< const CTP::instruments::CreditDefaultSwap >::deallocate);
    } else {
        env->SetLongField(arg0, JavaCPP_limitFID, rsize0 + position0);
    }
    const CTP::termstructures::IRates* rptr3 = (const CTP::termstructures::IRates*)adapter3;
    jlong rsize3 = (jlong)adapter3.size;
    void* rowner3 = adapter3.owner;
    if (rptr3 != (const CTP::termstructures::IRates*)ptr3) {
        JavaCPP_initPointer(env, arg3, rptr3, rsize3, rowner3, &SharedPtrAdapter< const CTP::termstructures::IRates >::deallocate);
    } else {
        env->SetLongField(arg3, JavaCPP_limitFID, rsize3 + position3);
    }
    if (exc != NULL) {
        env->Throw(exc);
    }
}
```

**Reply by saudet** [reply, marked as answer]
I see, we should be able to control that with the second element of the `@Cast` annotation, something like this here:
https://github.com/bytedeco/javacpp-presets/blob/1.5.8/pytorch/src/main/java/org/bytedeco/pytorch/presets/torch.java#L2499

**Reply by domcir** [reply]
I used `annotations` and it works (at least for parsing and compilation).

Thank you!!!
